### PR TITLE
Convert known_schemas to a GList

### PR DESCRIPTION
**Pull request by clumens**
Along the way, clean up and simplify a bunch of things.  This is in preparation for #3218.

### CODE REVIEW

File: lib/common/schemas.c [hunk 1]
```diff
@@ -177,63 +195,61 @@ add_schema(enum schema_validator_e validator, const schema_version_t *version,
...
 
     if (have_version) {
-        known_schemas[last].name = schema_strdup_printf("pacemaker-", *version, "");
+        schema->name = schema_strdup_printf("pacemaker-", *version, "");
     } else {
         CRM_ASSERT(name != NULL);
-        schema_scanf(name, "%*[^-]-", known_schemas[last].version, "");
-        known_schemas[last].name = strdup(name);
-        CRM_ASSERT(known_schemas[last].name != NULL);
+        schema_scanf(name, "%*[^-]-", schema->version, "");
```
Comment (clumens) [inline lib/common/schemas.c:224]: CI is failing because coverity is complaining about not checking scanf here.  This isn't new - it's just coming up now because the line changed.  It's not a problem to check it, but I'm not really sure what we'd do with the result.

Comment (clumens) [reply lib/common/schemas.c]: Added a patch on top to remove the unnecessary code here.

**Review by kgaillot — COMMENTED** [review]
Looks good

File: lib/common/schemas.c [hunk 2]
```diff
@@ -1260,3 +1260,24 @@ cli_config_update(xmlNode **xml, int *best_version, gboolean to_logs)
     free(orig_value);
     return rc;
 }
+
+void
+pcmk__dump_known_schemas(void)
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/common/schemas.c:1269]: Renamed.  For library size reasons, I wonder if it's worth hiding this behind some sort of compile-time define.

File: lib/common/schemas.c [hunk 3]
```diff
@@ -513,43 +513,60 @@ validate_with_relaxng(xmlDocPtr doc, xmlRelaxNGValidityErrorFunc error_handler,
...
+
+    free(schema->name);
+
+    if (schema->transform != NULL) {
+        free(schema->transform);
+    }
+
+    if (schema->transform_enter != NULL) {
+        free(schema->transform_enter);
+    }
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/common/schemas.c:556]: Fixed.

File: lib/common/schemas.c [hunk 4]
```diff
@@ -1074,41 +1074,48 @@ update_validation(xmlNode **xml_blob, int *best, int max, gboolean transform,
...
                        /* possibly avoid transforming when readily valid
                           (in general more restricted when crossing the major
                           version boundary, as X.0 "transitional" version is
                           expected to be more strict than it's successors that
                           may re-allow constructs from previous major line) */
-                       || validate_with_silent(xml, next >= 0 ? &known_schemas[next] : NULL)) {
+                       || validate_with_silent(xml, next_schema)) {
                 crm_debug("%s-style configuration is also valid for %s",
-                           schema->name, known_schemas[next].name);
+                           schema->name, next_schema->name);
```
[the remark this answers is not in the document]
Comment (clumens) [reply lib/common/schemas.c:1106]: next_schema?  Hm, no I don't think so.  The combination of setting it to `lpc+1` earlier in this block and then checking that it's less than max/max_stable_schemas should mean that it's always valid, unless the known_schemas list gets screwed up somehow.  I can add a check and a log message just in case, though.

This has me wondering if setting next to -1 in various places is still necessary, or if I've screwed something up.

[the remark this answers is not in the document]
Comment (clumens) [reply lib/common/schemas.c:1106]: Okay, the -1 stuff is definitely important and still has a use.  I added a `CRM_ASSERT(next_schema != NULL)` that never gets triggered by our tests, so I think we're okay to leave that in.

File: lib/common/schemas.c [hunk 5]
```diff
@@ -85,27 +76,54 @@ static int
...
+
+    /* Get the most recent schema so we can look at its version number. */
+    best = xml_latest_schema_index();
+    best_schema = g_list_nth(known_schemas, best)->data;
+
+    /* Iterate over the schema list until we find a schema with the same major
+     * version as best, and with a minor version number of 0.
+     *
+     * This assumes that the first schema in a major series is always X.0,
+     * which seems like a safe assumption.
```
Comment (kgaillot) [inline lib/common/schemas.c:104]: It should be, but we could avoid the assumption by stepping backward with prev() and returning the earliest schema with the same major version -- should be quicker in most cases too. You can do that in a separate PR if desired.

**Review by kgaillot — APPROVED** [review]

File: lib/common/schemas.c (hunk 5 above)
Comment (clumens) [reply lib/common/schemas.c:104]: See #3262.  This would be great to have a unit test for, but I don't want to make the function and the struct public.
